### PR TITLE
External Media: Fix Modal focus loss on arrow keypress

### PR DESCRIPTION
**Pull request by WunderBart**
Fixes: https://github.com/Automattic/jetpack/issues/15855

When the External Media modal is open, pressing any arrow key causes it to close immediately. 

To address this issue, I figured that a good thing to start with would be to handle the open/close toggle via the editor state (rather than re-rendering the whole thing every time content source changes). I've used other core implementations as an example - i.e.[ KeboardShortcutHelpModal](https://github.com/WordPress/gutenberg/blob/dbcaf38e886781d0bf5101db8cd04b5360ce2685/packages/edit-post/src/components/keyboard-shortcut-help-modal/index.js) or [OptionsModal](https://github.com/WordPress/gutenberg/blob/34ed2a6042d42fa18a5dcd0853d59bdff6a068d9/packages/edit-post/src/components/options-modal/index.js). I was hoping it would fix the issue, but that didn't happen, unfortunately. I'd still leave that change though, as it still improves the visibility toggle logic.

I've kept debugging and found out that `React` fires a `blur` event on the modal element right after the (arrow) `keydown` one. This happens probably during the re-render cycle of the Image block, which the modal component is a part of. The fun fact is that right after that, React actually tries to restore the focus to the modal element, but because the blur already happened, modal's [`withFocusOutside`](https://github.com/WordPress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/components/src/higher-order/with-focus-outside/index.js) handler thinks that [the outside of the modal was clicked](https://github.com/WordPress/gutenberg/blob/3b5bfa3964c336597687b62ac43906109c05f9aa/packages/components/src/modal/frame.js#L46) so it closes the modal anyway. So, I figured that we could fix that issue by completely isolating the modal from the Image block render scope. I was able to confirm this by triggering (via editor state) other (core) modals with the same button - when the modal component is outside the Image block render scope, blur is not being triggered, arrows are being contained and the focus stays within the modal. I've left [some commented out `openModal` calls](https://github.com/Automattic/jetpack/pull/16055/commits/2d6e89ec11a92a623d205548cf3b2db92bef19be#diff-b10daa55ee0c1b5a7218a11a0fe3b04bR34-R35) for testing purposes. To check it yourself just uncomment one of them and open the external media modal as you'd normally do. It will open the uncommented modal instead and you'll be able to see that arrow keys are being contained when pressed.

Unfortunately, I couldn't find a way to isolate the modal as mentioned above, so in https://github.com/Automattic/jetpack/pull/16055/commits/d9ec352830ff5f258f2f457905e50d755bc6396d I added a propagation handler for arrows `keydown` event. It makes sure that the `keydown` event doesn't propagate further, which fixes the https://github.com/Automattic/jetpack/issues/15855 while still keeping arrow keys functional inside the modal (i.e. when navigating inside Pexels search input text or choosing filter from one of the Google Photos filters' select dropdown). This solution is not ideal but it might be the best we can get now.

---

**Update:**

I've updated the code to include only the arrow keys handler. All the other changes were creating too many issues and weren't necessary ATM. This should be now easier to review and test.

### CODE REVIEW

File: extensions/shared/external-media/external-media-modal.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
+/**
```
Comment (WunderBart) [inline extensions/shared/external-media/external-media-modal.js:1]: FYI, this is a refactored version of `withMedia` HOC. 90% of the original code has been copied over. It's just a standalone component now and uses the editor state for the modal visibility toggle.

**Comment by obenland** [comment]
I took this for a test, the things that came to mind for me:
* Would you mind changing the base branch to `feature/external-media` and rebase it? There are changes to useMedia in there.
* After applying the patch I could get the modal to show up when replacing an image in a Media & text block.
* Is there no way we can keep `ExternalMediaModal` as a HOC?
* Could `handleArrowKeysPropagation()` be extended to prevent the upload modal to come up when adding images to a gallery block?

File: extensions/shared/external-media/external-media-modal.js [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+				}
+			} );
+
+			this.arrowKeysPropagationStopped = true;
+		}
+
+		if ( ! eventListenerEl && this.arrowKeysPropagationStopped ) {
+			this.arrowKeysPropagationStopped = false;
+		}
+	}
```
Comment (jeryj) [inline extensions/shared/external-media/external-media-modal.js:97]: A modal should catch _all_ keypresses within it. Nothing should propagate outside, but an escape keypress should close the modal. So, rather than catch specific keys (arrows) and exclude them, should we really just be trapping _all_ keys?

File: extensions/shared/external-media/external-media-modal.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+			isCopying: null,
+			requiresAuth: false,
+			path: { ID: PATH_RECENT },
+		};
+
+		this.arrowKeysPropagationStopped = false;
+	}
+
+	componentDidUpdate() {
+		this.handleArrowKeysPropagation();
```
Comment (jeryj) [inline extensions/shared/external-media/external-media-modal.js:60]: Why is this added on update instead of on mount?

File: extensions/shared/external-media/external-media-modal.js (hunk 2 above)
Comment (WunderBart) [reply extensions/shared/external-media/external-media-modal.js:97]: > A modal should catch all keypresses within it.

This would be true for our modal, but i.e. the `KeboardShortcutHelpModal` uses global key bindings for visibility toggle (`^⌥H` to open it from anywhere in the editor). So if all the keys were contained, that toggle would not work.

In our modal, escape is already handled and contained [here](https://github.com/WordPress/gutenberg/blob/30bc64c73daeae879f344992e9f74131d7366acb/packages/components/src/modal/frame.js#L84). If we stop the propagation of all keys, the [original  `keydown` handler of the modal](https://github.com/WordPress/gutenberg/blob/30bc64c73daeae879f344992e9f74131d7366acb/packages/components/src/modal/frame.js#L69) would not be reached, effectively disabling the escape key. We could instead block everything **except** escape. but I'm not sure if that's necessary? What if we block some global bindings too?

File: extensions/shared/external-media/external-media-modal.js (hunk 3 above)
Comment (WunderBart) [reply extensions/shared/external-media/external-media-modal.js:60]: The modal is mounted when the block is added to the editor canvas and unmounted when the block is removed. The listener element is not available on mount because it is rendered only when the modal is opened.

Comment (WunderBart) [reply extensions/shared/external-media/external-media-modal.js]: Decided to restore the original implementation - now the handler is attached on mount :)
See https://github.com/Automattic/jetpack/pull/16055/commits/bf3eb0f2bed2bdcfd8ac81f05c833715a6ac2b85.

**Comment by WunderBart** [comment]
I've updated (and rebased) the code to include only the arrow keys handler. All the other changes were creating too many issues and weren't necessary ATM. This should be now easier to review and test.

**Comment by jeryj** [comment]
@WunderBart and I went over this a lot today and isolated that this is not an issue specific to this modal. This happens to any Gutenberg `<Modal>` component used within the editor: https://github.com/WordPress/gutenberg/issues/22940

**Comment by WunderBart** [comment]
> @WunderBart and I went over this a lot today and isolated that this is not an issue specific to this modal. This happens to any Gutenberg `<Modal>` component used within the editor: [WordPress/gutenberg#22940](https://github.com/WordPress/gutenberg/issues/22940)

Nice find! We can probably still continue with this solution until a proper Gutenberg fix is landed, though. What do you think?

File: extensions/shared/external-media/sources/with-media.js [hunk 4]
```diff
@@ -53,6 +54,45 @@ export default function withMedia() {
...
+				 * When the External Media modal is open, pressing any arrow key causes
+				 * it to close immediately. This is happening because the keydown event
+				 * propagates outside the modal, triggering a re-render and a blur event
+				 * eventually. We could avoid that by isolating the modal from the Image
+				 * block render scope, but it is not possible in current implementation.
+				 *
+				 * This handler makes sure that the keydown event doesn't propagate further,
+				 * which fixes the issue described above while still keeping arrow keys
+				 * functional inside the modal.
+				 */
```
Comment (jeryj) [inline extensions/shared/external-media/sources/with-media.js:90]: Now that we've isolated this to a core modal issue, maybe we could mention the ongoing issue here? https://github.com/WordPress/gutenberg/issues/22940

Basically, how do we know to remove this code once it's fixed in the core Modal?

File: extensions/shared/external-media/sources/with-media.js [hunk 5]
```diff
@@ -53,6 +54,45 @@ export default function withMedia() {
...
+				const eventListenerEl = document.querySelector( '.components-modal__content' );
+				if ( ! eventListenerEl ) {
+					return null;
+				}
+
+				eventListenerEl.addEventListener( 'keydown', this.stopArrowKeysPropagation );
+				return () => {
+					eventListenerEl.removeEventListener( 'keydown', this.stopArrowKeysPropagation );
+				};
+			};
```
Comment (jeryj) [inline extensions/shared/external-media/sources/with-media.js:77]: It took me awhile to understand what was happening here. I think it would be easier to understand with:
- more comments (`attachArryKeysPropagationHandler` returning the `removeEventListener` for itself threw me for a minute)
- Restructuring the namings. Same as above, `this.removeArrowKeysPropagationHandler = this.attachArrowKeysPropagationHandler();` is a strange line. 

If this is an accepted way of going about this pattern, then we can leave it. It's my first time seeing this structure. It does work quite well though!

File: extensions/shared/external-media/sources/with-media.js [hunk 6]
```diff
@@ -53,6 +54,45 @@ export default function withMedia() {
...
+			}
+
+			componentWillUnmount() {
+				if ( this.removeArrowKeysPropagationHandler ) {
+					this.removeArrowKeysPropagationHandler();
+				}
+			}
+
+			attachArrowKeysPropagationHandler = () => {
+				const eventListenerEl = document.querySelector( '.components-modal__content' );
```
Comment (jeryj) [inline extensions/shared/external-media/sources/with-media.js:68]: Is there a `ref` or something we can utilize instead, or is this the only way to get the `el`?

Comment (WunderBart) [reply extensions/shared/external-media/sources/with-media.js:68]: There's no ref forwarded to the Modal, but that's a good point and I'm going to check if we can use our own content wrapper to attach the listener, so we don't rely on external class name.

File: extensions/shared/external-media/sources/with-media.js (hunk 5 above)
Comment (WunderBart) [reply extensions/shared/external-media/sources/with-media.js:77]: That's a convention that's used i.e. [in the `useEffect` hook](https://reactjs.org/docs/hooks-effect.html#effects-without-cleanup), where you can return a `cleanup` function that will be executed when the component unmounts. I don't think there's a name for it, though, but basically, it's a subscriber function that returns its own unsubscriber. This is also a good example:

```js
const observable = new Observable(function subscribe(subscriber) {
  // Keep track of the interval resource
  const intervalId = setInterval(() => {
    subscriber.next('hi');
  }, 1000);

  // Provide a way of canceling and disposing the interval resource
  return function unsubscribe() {
    clearInterval(intervalId);
  };
});
```

Source: https://rxjs.dev/guide/observable#disposing-observable-executions

Comment (jeryj) [reply extensions/shared/external-media/sources/with-media.js:77]: Could we do something similar to how they have it set-up in the Observable example?

`attachArrowKeysPropagationHandler` -> `arrowKeysPropagationHandler`

`return () => ` -> `return remove() => `

So, something like:
```
componentDidMount() {
  this.removeArrowKeysPropagationHandler = this.arrowKeysPropagationHandler.remove();
}

arrowKeysPropagationHandler = () => {
  // ... handle adding the listener

  // give the anonymous function a name
  return remove() => {
    ...
  }
}
```

**Comment by WunderBart** [comment]
@jeryj, this is ready for another go! 🤞 I was able to make a lot simpler after your comments. Thanks! 

FYI so you don't miss it - [I removed the `onMouseDown` event stopper](https://github.com/Automattic/jetpack/pull/16055/commits/ae77fc44d9efd3501c2245946a308f87bcd259c3) as it's already implemented in the current Modal implementation!

File: extensions/shared/external-media/sources/with-media.js (hunk 5 above)
Comment (WunderBart) [reply extensions/shared/external-media/sources/with-media.js]: No need to use lifecycle calls actually :D Attached it directly here: https://github.com/Automattic/jetpack/pull/16055/commits/64f43198c7cca9fdcd4f726596da9bc35526ce04

File: extensions/shared/external-media/sources/with-media.js (hunk 4 above)
Comment (WunderBart) [reply extensions/shared/external-media/sources/with-media.js]: Added in https://github.com/Automattic/jetpack/pull/16055/commits/7337df3b19d2b86dd9d0bb235f38a8886544299f

**Comment by WunderBart** [comment]
> If you move focus to outside the modal content and press the arrow key, it will still close the modal:

Ugh, how did I miss that! I think we need to re-attach the handler to the modal listener element then. Back to the more complicated version! 😬

File: extensions/shared/external-media/sources/with-media.js [hunk 7]
```diff
@@ -53,6 +54,50 @@ export default function withMedia() {
...
+				if ( this.arrowKeysPropagationHandler ) {
+					this.arrowKeysPropagationHandler.remove();
+				}
+			}
+
+			attachArrowKeysPropagationHandler = () => {
+				const eventListenerEl = document.querySelector( '.jetpack-external-media-browser' );
+				if ( ! eventListenerEl ) {
+					return null;
+				}
```
Comment (marekhrabe) [inline extensions/shared/external-media/sources/with-media.js:71]: This is dangerous, since the whole document is queried for the element, not just this component. We should instead use `ref` on the target element and add/remove listensers inside the `ref` handler.

Comment (WunderBart) [reply extensions/shared/external-media/sources/with-media.js:71]: I wanted to use ref, but can't really see how we could do that since the Modal component doesn't allow for ref passing to the container/listener element. FYI, this won't work [when attached to the content wrapper](https://github.com/Automattic/jetpack/pull/16055/commits/ab9d7afd36ed41fbaf71ebbcc937139fe8dde701), because of [this](https://github.com/Automattic/jetpack/pull/16055#pullrequestreview-426551585). Any idea if we still can use a ref? If we can't, what would be the danger of querying that element?

Comment (marekhrabe) [reply extensions/shared/external-media/sources/with-media.js]: Gotcha! Since it wasn't possible to attach `ref` directly to the modal, I've tried adding a new element _within_ the modal and walk our way up though the DOM to get the modal element.

Comment (WunderBart) [reply extensions/shared/external-media/sources/with-media.js]: Thanks! Just tested it locally - works like a charm!

File: extensions/shared/external-media/sources/with-media.js [hunk 8]
```diff
@@ -53,6 +54,42 @@ export default function withMedia() {
...
+					this.modalElement = el.closest( '.jetpack-external-media-browser' );
+
+					// Attach the listener if found.
+					if ( this.modalElement ) {
+						this.modalElement.addEventListener( 'keydown', this.stopArrowKeysPropagation );
+					}
+				} else if ( this.modalElement ) {
+					// Remove listeners when unmounting.
+					this.modalElement.removeEventListener( 'keydown', this.stopArrowKeysPropagation );
+					this.modalElement = null;
```
Comment (jeryj) [inline extensions/shared/external-media/sources/with-media.js:69]: Can you explain how this works? Does `modalRef` get called again on unmount? And at that point, the `el` would be undefined since it's unmounted, but `this.modalElement` would still exist, thus the event listener would get removed?

Comment (jeryj) [reply extensions/shared/external-media/sources/with-media.js:69]: Confirmed this behavior when testing. Works well!

Comment (marekhrabe) [reply extensions/shared/external-media/sources/with-media.js:69]: You got it. Ref gets called after the DOM element gets mounted and the element is passed as an argument. The next time it is called, it is just before unmounting with an empty argument that tells you to clean up.
